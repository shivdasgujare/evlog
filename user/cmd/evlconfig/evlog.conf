#
# Configuration values for evlogd daemon. 
# Use 'evlconfig' command to update these values. If this file is edited 
# manually, the updated value will not be used by the evlogd daemon. So,
# the evlogd daemon has to be restarted. Hence, use 'evlconfig' command
# to send this values to evlogd daemon.
# Console display level - OFF=111
#
Discard Dups: on
Discard Interval: 3
Discard Count: 100
Event Screen: nofilter
Console display level: 111
Lookback Size: 3
