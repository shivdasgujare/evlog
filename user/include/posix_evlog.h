/*
 * Linux Event Logging for the Enterprise
 * Copyright (c) International Business Machines Corp., 2001
 *
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 *  Please send e-mail to lkessler@users.sourceforge.net if you have
 *  questions or comments.
 *
 *  Project Website:  http://evlog.sourceforge.net/
 *
 */

#define LOG_LOGMGMT (12<<3) /* this facility defined by POSIX, but not syslog */
#ifndef _POSIX_EVLOG_H_
#define _POSIX_EVLOG_H_

#include <sys/types.h>
/* For sigevent */
#include <signal.h>
/* For facility (e.g., LOG_KERN) and severity (e.g., LOG_EMERG) codes */
#include <syslog.h>
/* For va_list */
#include <stdarg.h>

#ifdef __cplusplus
extern "C" {
#endif	/* __cplusplus */

#ifdef _POSIX_THREADS
#include <pthread.h>
#endif
#include <linux/evl_log.h>

/* Treat a log descriptor like a file descriptor for now. */
typedef struct _log_description *posix_logd_t;

/* Member selectors for posix_log_entry, for use by posix_log_memtostr  */
#define POSIX_LOG_ENTRY_DATA	0	/* variable-length data */
#define POSIX_LOG_ENTRY_RECID	1
#define POSIX_LOG_ENTRY_SIZE	2
#define POSIX_LOG_ENTRY_FORMAT	3
#define POSIX_LOG_ENTRY_EVENT_TYPE	4
#define POSIX_LOG_ENTRY_FACILITY	5
#define POSIX_LOG_ENTRY_SEVERITY	6
#define POSIX_LOG_ENTRY_UID		7
#define POSIX_LOG_ENTRY_GID		8
#define POSIX_LOG_ENTRY_PID		9
#define POSIX_LOG_ENTRY_PGRP	10
#define POSIX_LOG_ENTRY_TIME	11
#define POSIX_LOG_ENTRY_FLAGS	12
#define POSIX_LOG_ENTRY_THREAD	13
#define POSIX_LOG_ENTRY_PROCESSOR	14

/* Query purpose flags */
#define POSIX_LOG_PRPS_NOTIFY	0x1
#define POSIX_LOG_PRPS_GENERAL	0x2
#define POSIX_LOG_PRPS_SEEK	(POSIX_LOG_PRPS_GENERAL)
#define EVL_PRPS_TEMPLATE	0x4
#define EVL_PRPS_RESTRICTED	0x8
#define POSIX_LOG_PRPS_MASK	0xf

struct node;

typedef struct _posix_log_query {
	int		qu_purpose;	/* NOTIFY, SEEK, etc. */
	char		*qu_expr;	/* Saved copy of the query expr */
	struct node	*qu_tree;	/* Parse tree */
	struct evl_nonStdAtts	*qu_nonStdAtts;	/* Non-standard attributes. */
} posix_log_query_t;


#define POSIX_LOG_ONCE_ONLY		0x01
#define POSIX_LOG_SEND_RECID		0x02
#define POSIX_LOG_SEND_SIGVAL		0x04
#define POSIX_LOG_NFY_MASK		0x07

#define POSIX_LOG_NFY_DISABLED		0x08
#define POSIX_LOG_ACTION_PERSIST	0x16

#define POSIX_LOG_NFY_PENDING_REMOVE	0x512

#define SI_EVLOG	SI_QUEUE
typedef size_t posix_log_notify_t;
/*
 * These are the types of messages that can be sent between the client (a
 * process that has called posix_log_notify_add_ex()) and the server (the
 * notification daemon).
 */
typedef enum _nfyMsgType {
	nmtNewRequest,			/* new notification request */
	nmtGetRequestStatus,	/* get status of notification request */
	nmtRmRequest,			/* remove notification request */
	nmtRequestStatus		/* response to client requests */
} notifyMsgType_t;

/*
 * The status of a notification request.
 */
typedef enum _nfyRqStatus {
	nrsEnabled,		/* Request has been accepted by server. */
	nrsDisabled,	/* Once-only request has been satisfied */
					/* by the server, but not removed by the */
					/* client. */
	nrsRemoved,		/* Request has been removed at client's request. */
	nrsNoRequest	/* Request not found by server */
} nfyRqStatus_t;

/*
 * The header of a message between client and server.
 */
typedef struct _nfyMsgHdr {
	int	nhMsgType;
	posix_log_notify_t	nhReqHandle;
	int	nhStatus;
} nfyMsgHdr_t;

typedef struct _nfyNewRqMsg {
	nfyMsgHdr_t	nwHeader;
	size_t		nwQueryLength;
	struct sigevent	nwSigevent;
	int		nwFlags;
} nfyNewRqMsg_t;

#define EVL_NFY_MSGMAXSIZE (sizeof(nfyNewRqMsg_t))

/* Log-management event-type codes */
#define POSIX_LOG_MGMT_TIMEMARK	1
#define POSIX_LOG_MGMT_STARTMAINT	2
#define POSIX_LOG_MGMT_ENDMAINT	3
#define POSIX_LOG_MGMT_STOPLOGGING	4
#define POSIX_LOG_MGMT_STARTLOGGING	5

/* Per-process maximum number of concurrent open log descriptors */
#define POSIX_LOG_OPEN_MAX	100

/* Per-process max number of concurrently registered notification requests */
#define POSIX_LOG_NOTIFY_MAX	100

/* Seek directions */
#define POSIX_LOG_SEEK_START	1
#define POSIX_LOG_SEEK_END		2
#define POSIX_LOG_SEEK_FORWARD	3
#define POSIX_LOG_SEEK_BACKWARD	4
#define POSIX_LOG_SEEK_FIRST	5
#define POSIX_LOG_SEEK_LAST		6

/* Maximum length for a string returned by posix_log_memtostr */
#define POSIX_LOG_MEMSTR_MAXLEN	128

/* Log an event */
extern int posix_log_write(posix_log_facility_t facility,
	int event_type, posix_log_severity_t severity,
	const void *buf, size_t len, int format, unsigned int flags);

extern int posix_log_vprintf(posix_log_facility_t facility,
	int event_type, posix_log_severity_t severity, unsigned int flags,
	const char *format, va_list args);

extern int posix_log_printf(posix_log_facility_t facility,
	int event_type, posix_log_severity_t severity, unsigned int flags,
	const char *format, ...);

extern int posix_log_vprintb(posix_log_facility_t facility,
	int event_type, posix_log_severity_t severity, unsigned int flags,
	const char *format, va_list args);

extern int posix_log_printb(posix_log_facility_t facility,
	int event_type, posix_log_severity_t severity, unsigned int flags,
	const char *format, ...);

/* Open an event log for read access */
extern int posix_log_open(posix_logd_t *logdes, const char *path);

/* Read from event log */
extern int posix_log_read(posix_logd_t logdes, struct posix_log_entry *entry,
	void *log_buf, size_t log_len); 

/* Register for notification about new events */
extern int posix_log_notify_add(const posix_log_query_t *query,
	const struct sigevent *notification, int flags,
	posix_log_notify_t *nfyhandle);

extern int posix_log_sigval_getrecid(const union sigval sval,
	posix_log_recid_t *recid);

extern int posix_log_siginfo_getrecid(const siginfo_t *info,
        void *context, posix_log_recid_t *recid);
        
extern int posix_log_notify_get(posix_log_notify_t nfyhandle,
	struct sigevent *notification, int *flags,
	char *qsbuf, size_t qslen, size_t *reqlen);

/* Remove notification request */
extern int posix_log_notify_remove(posix_log_notify_t nfyhandle);

/* Close event log */
extern int posix_log_close(posix_logd_t logdes);

/* Reposition the read pointer */
extern int posix_log_seek(posix_logd_t logdes, const posix_log_query_t *query,
	int direction); 

/* Compare severities */
extern int posix_log_severity_compare(int *order, posix_log_severity_t s1,
	posix_log_severity_t s2); 

/* Create log query */
extern int posix_log_query_create(const char *query_string, int purpose,
	posix_log_query_t *query, char *errbuf, size_t errlen);

extern int posix_log_query_get(const posix_log_query_t *query, int *purpose,
	char *qsbuf, size_t qslen, size_t *reqlen);

/* Destroy log query */
extern int posix_log_query_destroy(posix_log_query_t *query);

/* Obtain the string equivalent of an event-record member */
extern int posix_log_memtostr(int member_selector,
	const struct posix_log_entry *entry,
	char *buf, size_t buflen);

/* Convert string to facility or facility to string */
extern int posix_log_factostr(posix_log_facility_t fac, char *buf,
	size_t buflen);

extern int posix_log_strtofac(const char *str, posix_log_facility_t *fac);

#ifdef __cplusplus
}
#endif 	/* __cplusplus */
#endif	/* _POSIX_EVLOG_H_ */
