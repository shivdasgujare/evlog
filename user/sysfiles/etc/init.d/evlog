#!/bin/sh
#
# chkconfig: 35 09 90
# description: Enterprise event log daemon
#
#

# Source function library.

REDHAT=`cat /proc/version | grep "Red Hat"`
# Source function library.
if [ ! -z "$REDHAT" ]
then
	. /etc/init.d/functions
fi


[ -f /sbin/evlogd ] || exit 0

EVLOGD_OPTIONS=""
# Source config
if [ -f /etc/sysconfig/evlog ]; then
	. /etc/sysconfig/evlog
fi

RETVAL=0

umask 077

record_last_recid() {
	/sbin/evlview -l $1 -t 1 -S %recid% > $2 2> /dev/null
	test -s $2 || echo 0 > $2
	chmod 644 $2
}

start() {
 	echo -n $"Starting enterprise event logger: "

	record_last_recid /var/evlog/eventlog /var/evlog/bootrecid
	record_last_recid /var/evlog/privatelog /var/evlog/pbootrecid
 	
 	if [ ! -z "$REDHAT" ]
	then 	
		daemon /sbin/evlogd $EVLOGD_OPTIONS
	else
		/sbin/evlogd $EVLOGD_OPTIONS
	fi
	sleep 1
	RETVAL=$?
	echo
	touch /var/lock/subsys/evlog
	return $RETVAL
}	
stop() {
	echo -n $"Shutting down enterprise event logger: "
	
	if [ ! -z "$REDHAT" ]
	then
		killproc evlogd
	else
		PID=`ps -ef | grep evlogd | grep -v grep | awk '$3==1 {print $2}'`
		if [ ! -z "$PID" ]
		then
			kill -s SIGTERM $PID
		fi
	fi

	RETVAL=$?
	echo
	rm -f /var/lock/subsys/evlog
	if [ -e /var/run/evlogd.pid ]
	then
		rm -f /var/run/evlogd.pid
	fi
	return $RETVAL
}
rhstatus() {
	status evlogd
}
restart() {
	stop
	start
}	

case "$1" in
  start)
  	start
	;;
  stop)
  	stop
	;;
  status)
  	rhstatus
	;;
  restart|reload)
  	restart
	;;
  condrestart)
    restart	
	;;
  *)
	echo $"Usage: $0 {start|stop|status|restart|condrestart}"
	exit 1
esac

exit $?

