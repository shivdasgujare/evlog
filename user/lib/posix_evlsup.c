/*
 * Linux Event Logging for the Enterprise
 * Copyright (c) International Business Machines Corp., 2001
 *
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 *  Please send e-mail to lkessler@users.sourceforge.net if you have
 *  questions or comments.
 *
 *  Project Website:  http://evlog.sourceforge.net/
 *
 */
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <ctype.h>
#include <sys/syslog.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <fcntl.h>
#include <assert.h>
#include <stdlib.h>

#include "config.h"

#include "posix_evlog.h"
#include "posix_evlsup.h"
#include "evlog.h"

/*
 * These are the facilities defined by Linux's sys/syslog.h.
 * The facility registry must always contain at least these.
 */
struct _evlNvPair _evlLinuxFacilities[] = {
	{ LOG_AUTH,	"AUTH" },
	{ LOG_CRON,	"CRON" },
	{ LOG_DAEMON,	"DAEMON" },
	{ LOG_KERN,	"KERN" },
	{ LOG_LPR,	"LPR" },
	{ LOG_MAIL,	"MAIL" },
	{ LOG_NEWS,	"NEWS" },
	{ LOG_SYSLOG,	"SYSLOG" },
	{ LOG_USER,	"USER" },
	{ LOG_LOGMGMT,	"LOGMGMT" },

	{ LOG_UUCP,	"UUCP" },
	{ LOG_AUTHPRIV,	"AUTHPRIV" },
	{ LOG_FTP,	"FTP" },

	{ LOG_LOCAL0,	"LOCAL0" },
	{ LOG_LOCAL1,	"LOCAL1" },
	{ LOG_LOCAL2,	"LOCAL2" },
	{ LOG_LOCAL3,	"LOCAL3" },
	{ LOG_LOCAL4,	"LOCAL4" },
	{ LOG_LOCAL5,	"LOCAL5" },
	{ LOG_LOCAL6,	"LOCAL6" },
	{ LOG_LOCAL7,	"LOCAL7" },
	{ 0, 0 }
};

/*
 * These are the facilities defined by POSIX 1003.25.
 */
struct _evlNvPair _evlPosixFacilities[] = {
	{ LOG_AUTH,	"AUTH" },
	{ LOG_CRON,	"CRON" },
	{ LOG_DAEMON,	"DAEMON" },
	{ LOG_KERN,	"KERN" },
	{ LOG_LPR,	"LPR" },
	{ LOG_MAIL,	"MAIL" },
	{ LOG_NEWS,	"NEWS" },
	{ LOG_SYSLOG,	"SYSLOG" },
	{ LOG_USER,	"USER" },
	{ LOG_LOGMGMT,	"LOGMGMT" },
	{ LOG_UUCP,	"UUCP" },
	{ LOG_LOCAL0,	"LOCAL0" },
	{ LOG_LOCAL1,	"LOCAL1" },
	{ LOG_LOCAL2,	"LOCAL2" },
	{ LOG_LOCAL3,	"LOCAL3" },
	{ LOG_LOCAL4,	"LOCAL4" },
	{ LOG_LOCAL5,	"LOCAL5" },
	{ LOG_LOCAL6,	"LOCAL6" },
	{ LOG_LOCAL7,	"LOCAL7" },
	{ 0, 0 }
};

struct _evlNvPair _evlSeverities[] = {
	{ LOG_EMERG,	"EMERG" },
	{ LOG_ALERT,	"ALERT" },
	{ LOG_CRIT,	"CRIT" },
	{ LOG_ERR,	"ERR" },
	{ LOG_WARNING,	"WARNING" },
	{ LOG_NOTICE,	"NOTICE" },
	{ LOG_INFO,	"INFO" },
	{ LOG_DEBUG,	"DEBUG" },
	{ 0, 0 }
};

struct _evlNvPair _evlAttributes[] = {
	{ POSIX_LOG_ENTRY_RECID,	"recid" },
	{ POSIX_LOG_ENTRY_SIZE,	"size" },
	{ POSIX_LOG_ENTRY_FORMAT,	"format" },
	{ POSIX_LOG_ENTRY_EVENT_TYPE,	"event_type" },
	{ POSIX_LOG_ENTRY_FACILITY,	"facility" },
	{ POSIX_LOG_ENTRY_SEVERITY,	"severity" },
	{ POSIX_LOG_ENTRY_UID,	"uid" },
	{ POSIX_LOG_ENTRY_GID,	"gid" },
	{ POSIX_LOG_ENTRY_PID,	"pid" },
	{ POSIX_LOG_ENTRY_PGRP,	"pgrp" },
	{ POSIX_LOG_ENTRY_TIME,	"time" },
	{ POSIX_LOG_ENTRY_FLAGS,	"flags" },
	{ POSIX_LOG_ENTRY_THREAD,	"thread" },
	{ POSIX_LOG_ENTRY_PROCESSOR,	"processor" },
	{ 0, 0 }
};

struct _evlNvPair _evlFormats[] = {
	{ POSIX_LOG_NODATA,	"NODATA" },
	{ POSIX_LOG_BINARY,	"BINARY" },
	{ POSIX_LOG_STRING,	"STRING" },
	{ POSIX_LOG_PRINTF,	"PRINTF" },
	{ 0, 0 }
};

struct _evlNvPair _evlMgmtEventTypes[] = {
	{ POSIX_LOG_MGMT_TIMEMARK,		"MGMT_TIMEMARK" },
	{ POSIX_LOG_MGMT_STARTMAINT,	"MGMT_STARTMAINT" },
	{ POSIX_LOG_MGMT_ENDMAINT,		"MGMT_ENDMAINT" },
	{ POSIX_LOG_MGMT_STOPLOGGING,	"MGMT_STOPLOGGING" },
	{ POSIX_LOG_MGMT_STARTLOGGING,	"MGMT_STARTLOGGING" },
	{ 0, 0 }
};

void
_evlGetNameByValue(const struct _evlNvPair table[], int value, char *name, size_t size,
	const char *dflt)
{
	const struct _evlNvPair *nv;
	for (nv = table; nv->nv_name; nv++) {
		if (nv->nv_value == value) {
			(void) strcpy(name, nv->nv_name);
			return;
		}
	}

	/* No match */
	if (dflt) {
		(void) strcpy(name, dflt);
	} else {
		snprintf(name, size, "%d", value);
	}
}

/* Look up value by name, case-sensitive */
int
_evlGetValueByName(const struct _evlNvPair table[], const char *name, int dflt)
{
	const struct _evlNvPair *nv;
	for (nv = table; nv->nv_name; nv++) {
		if (!strcmp(nv->nv_name, name)) {
			return nv->nv_value;
		}
	}

	/* No match */
	return dflt;
}

/* Like strcmp, but case-insensitive */
int
_evlCIStrcmp(const char *s1, const char *s2)
{
	while (*s1 && toupper(*s1) == toupper(*s2)) {
		s1++;
		s2++;
	}
	return toupper(*s1) - toupper(*s2);
}

/* Look up value by name, case-insensitive */
int
_evlGetValueByCIName(const struct _evlNvPair table[], const char *name, int dflt)
{
	const struct _evlNvPair *nv;
	for (nv = table; nv->nv_name; nv++) {
		if (!_evlCIStrcmp(nv->nv_name, name)) {
			return nv->nv_value;
		}
	}

	/* No match */
	return dflt;
}

/* Return the proc id */
int 
_evlGetProcId()
{
	int fd, ret;
	char buf[1024];
	char *last_token;
	size_t r;
	
	if ((fd = open("/proc/self/stat", O_RDONLY)) == -1) {
		fprintf(stderr, "Can't open stat\n");
		exit(1);
	}

	if ((r = read(fd, buf, 1024)) == -1) {
		close(fd);
		fprintf(stderr, "Read stat failed\n");
		exit(1);
	}
	buf[r]='\0';
	close(fd);
	/* The last token contains the processor id */

	last_token = (char *) strrchr(buf, ' ');
	assert(last_token != NULL);	
	return (int) strtol(last_token, (char **)NULL, 10);
}
