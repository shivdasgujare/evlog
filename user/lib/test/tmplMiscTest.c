/*
 * IBM Event Logging for Linux
 * Copyright (c) International Business Machines Corp., 2001
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or 
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  
 */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include "posix_evlog.h"
#include "posix_evlsup.h"
#include "evlog.h"

/*
 * Logs event records to be interpreted using the "tmplMiscTest.t" template
 * of facility "user".
 */

#define LOGERR(evtype, sev, msg, args...) evl_log_write(LOG_USER, evtype, sev, 0, "string", msg, ##args)

int exitStatus = 0;

static void
checkStatus(int status, int evty)
{
	if (status != 0) {
		fprintf(stderr, "write of event type %d failed:\n", evty);
		errno = status;
		perror("evl_log_write");
		exitStatus = 1;
	}
}

struct familyFlags {
	unsigned int	fatherFlag:1;
	unsigned int	motherFlag:1;
	unsigned int	sonFlag:1;
	unsigned int	daughterFlag:1;
	unsigned int	catFlag:1;
	unsigned int	dogFlag:1;
	unsigned int	computerFlag:1;
	unsigned int	otherFlag:1;
};

main(int argc, char **argv)
{
	int status;
	int type;
	posix_log_severity_t sev = LOG_INFO;
	struct familyFlags ff;
	const char *msg;
	
	type = 5001;
	status = LOGERR(type, sev,
		"This is the message for event type 5001.",
		"endofdata");
	checkStatus(status, type);
	
#define human		0x1
#define inhuman		0x0
#define adult		0x2
#define juvenile	0x0
#define female		0x4
#define male		0x0
#define encodeOrganism(name,type) "string",name,"3*int",type,type,type
	type = 5012;
	status = LOGERR(type, sev,
		"This is the message for event type 5002.",
		encodeOrganism("Donald Duck", inhuman|male|adult),
		encodeOrganism("Daisy Duck", inhuman|female|adult),
		encodeOrganism("Dewey Duck", inhuman|male|juvenile),
		encodeOrganism("April Duck", inhuman|female|juvenile),
		encodeOrganism("Fred Flintstone", human|male|adult),
		encodeOrganism("Wilma Flintstone", human|female|adult),
		encodeOrganism("Pebbles Flintstone", human|female|juvenile),
		encodeOrganism("Bambam Rubble", human|male|juvenile),
		"endofdata");
	checkStatus(status, type);

	type = 5003;
	status = LOGERR(type, sev,
		"This is the message for event type 5003.",
		"string", "George W. Bush",
		"string", "1600 Pennsylvania Ave. NW",
		"string", "Washington",
		"string", "DC",
		"string", "20502",
		"string", "US President",
		"double", 200000.00,
		"2*char", 1, 20, "short", 2001,
		"endofdata");
	checkStatus(status, type);

	type = 5004;
	status = LOGERR(type, sev,
		"The Flagstons:",
		"char", 2, "2*string", "Hi", "Lois",
		"char", 2, "2*string", "Chip", "Ditto",
		"char", 2, "2*string", "Dot", "Trixie",
		"char", 0,
		"char", 1, "string", "Dawg",
		"char", 0,
		"char", 0,
		"endofdata");
	checkStatus(status, type);

	type = 5005;
	memset(&ff, 0, sizeof(ff));
	ff.fatherFlag = 1;
	ff.motherFlag = 1;
	ff.daughterFlag = 1;
	ff.otherFlag = 1;
	status = LOGERR(type, sev,
		"The Flintstones:",
		"char[]", sizeof(ff), &ff,
		"4*string", "Fred", "Wilma", "Pebbles", "Dino",
		"endofdata");
	checkStatus(status, type);

	type = 5006;
	status = LOGERR(type, sev,
		"This message should be followed by 3 extra strings.",
		"3*string", "Extra string 1", "Extra string 2", "Extra string 3",
		"endofdata");
	checkStatus(status, type);

	msg = "Test of specifying event_type "
		"as string in template.";
	type = (int) _evlCrc32((const unsigned char *)msg, strlen(msg));
	status = posix_log_printf(LOG_USER, type, sev, 0, msg);
	if (status != 0) {
		fprintf(stderr, "write of event type %#x failed:\n", type);
		errno = status;
		perror("posix_log_printf");
		exitStatus = 1;
	}

	exit(exitStatus);
}
