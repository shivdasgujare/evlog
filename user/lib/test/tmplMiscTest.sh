#!/bin/bash

export LANG=en_US
TEST_TEMPLATE=tmplMiscTest.out

EVLTMPLPATH=$EVL_TESTDIR/templates
export EVLTMPLPATH

if [ "$USER" = root ]
then
	/sbin/evltc $EVLTMPLPATH/hr/hr.t
	/sbin/evltc $EVLTMPLPATH/user/tmplMiscTest.t
else
echo >&2 "tmplMiscTest.sh: Not running as root.  Assuming the following"
echo >&2 "template source files jave already been compiled:"
echo >&2 $EVLTMPLPATH/hr/hr.t
echo >&2 $EVLTMPLPATH/user/tmplMiscTest.t
fi

logtime=`date +%s`
sleep 1
./tmplMiscTest
sleep 1
/sbin/evlview -f "time>${logtime}" -S 'event type %event_type:d%:\n%data%' > tmplMiscTest.sh.out

diff tmplMiscTest.sh.out $TEST_TEMPLATE

if [ $? -eq "0" ]
then
	echo templates misc test	:PASSED
else
	echo templates misc test	:FAILED
fi
