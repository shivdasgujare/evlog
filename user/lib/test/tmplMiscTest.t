facility "USER"; event_type 5001;
/* Test of various kinds of constants */
const {
	char	c = 'c' "%c";
	int	i = 1;
	double	d = 1.1;
	double	d2 = 2.2e2;
	string	s = "string s";

}
attributes {
	string	msg;
}
format
c=%c%
i=%i%
d=%d:.2f%
d2=%d2:.2f%
s=%s%
msg=%msg%
END

/* Test of %b and %v formats. */
struct _Organism;
attributes {
	string		name;

	int btype	"%b/b1/human/b0/inhuman/"
			"b1x/adult/b0x/juvenile/"
			"b1xx/female/b0xx/male/";

	int xtype	"%b/x7/Woman/"
			"x5/Girl/"
			"x1/Human/x2/Adult/x4/Female/";

	int vtype	"%v/1/boy/"
			"3/man/"
			"5/girl/"
			"7/woman/";
}
format
name=%name%, type=%btype% = %xtype% = %vtype%
END

typedef struct _Organism Organism;

facility "USER"; event_type 5012;
attributes {
	string		msg;
	Organism	folks[_R_]	"(%Z\n)";
}
format
%folks%
%msg%
END

/* Test of import dir.* */
import hr.*;

facility "USER"; event_type 5003;
description "This template uses the hr.employee template.";
attributes {
	string		msg;
	struct employee	employee;
}
format
%employee%
%msg%
END

facility "USER"; event_type 5004;
/* Test of the use of one attribute as another attribute's dimension */
attributes {
	string	msg;
	char	nParents;
	string	parents[nParents]	"(%s )";
	char	nSons;
	string	sons[nSons]		"(%s )";
	char	nDaughters;
	string	daughters[nDaughters]	"(%s )";
	char	nCats;
	string	cats[nCats]		"(%s )";
	char	nDogs;
	string	dogs[nDogs]		"(%s )";
	char	nComputers;
	string	computers[nComputers]	"(%s )";
	char	nOthers;
	string	others[nOthers]		"(%s )";
}
format
%msg%
parents: %parents%
sons: %sons%
daughters: %daughters%
cats: %cats%
dogs: %dogs%
computers: %computers%
others: %others%
END

struct familyFlags;
aligned attributes {
		uint	fatherFlag:1;
		uint	motherFlag:1;
		uint	sonFlag:1;
		uint	daughterFlag:1;
		uint	catFlag:1;
		uint	dogFlag:1;
		uint	computerFlag:1;
		uint	otherFlag:1;
}
format
fatherFlag: %fatherFlag%
motherFlag: %motherFlag%
sonFlag: %sonFlag%
daughterFlag: %daughterFlag%
catFlag: %catFlag%
dogFlag: %dogFlag%
computerFlag: %computerFlag%
otherFlag: %otherFlag%
END

facility "USER"; event_type 5005;
/*
 * Test of the use of one attribute as another attribute's dimension.
 * In this case, the dimension attribute is a member of a struct attribute
 * (and also a bit-field).
 */
attributes {
	string	msg;
	struct familyFlags f;
	string father[f.fatherFlag]	"(%s )";
	string mother[f.motherFlag]	"(%s )";
	string son[f.sonFlag]		"(%s )";
	string daughter[f.daughterFlag]	"(%s )";
	string cat[f.catFlag]		"(%s )";
	string dog[f.dogFlag]		"(%s )";
	string computer[f.computerFlag]	"(%s )";
	string others[f.otherFlag]	"(%s )";
}
format
%msg%
%f%
father: %father%
mother: %mother%
son: %son%
daughter: %daughter%
cat: %cat%
dog: %dog%
computer: %computer%
others: %others%
END

facility "USER"; event_type 5006;
/* Test of [*] as a synonym for [_R_] */
attributes {
	string msg;
	string extraMsgs[*]	"(%s\n)";
}
format
%msg%
%extraMsgs%
END

facility "USER";
event_type "Test of specifying event_type "
	"as string in template.";
attributes {
	string msg;
}
format
%msg%
If you see this line, we probably passed.
END
