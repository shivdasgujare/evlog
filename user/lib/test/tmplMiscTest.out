event type 5001:
c=c
i=1
d=1.10
d2=220.00
s=string s
msg=This is the message for event type 5001.
event type 5012:
name=Donald Duck, type=0x2(inhuman|adult|male) = 0x2(Adult) = 2
name=Daisy Duck, type=0x6(inhuman|adult|female) = 0x6(Adult|Female) = 6
name=Dewey Duck, type=0x0(inhuman|juvenile|male) = 0x0 = 0
name=April Duck, type=0x4(inhuman|juvenile|female) = 0x4(Female) = 4
name=Fred Flintstone, type=0x3(human|adult|male) = 0x3(Human|Adult) = man
name=Wilma Flintstone, type=0x7(human|adult|female) = 0x7(Woman) = woman
name=Pebbles Flintstone, type=0x5(human|juvenile|female) = 0x5(Girl) = girl
name=Bambam Rubble, type=0x1(human|juvenile|male) = 0x1(Human) = boy

This is the message for event type 5002.
event type 5003:
Employee Name:	George W. Bush
Home Address:
1600 Pennsylvania Ave. NW
Washington, DC 20502
USA
Job Title:	US President
Annual Salary:	200000.00
Start Date (m/d/y):	1/20/2001
This is the message for event type 5003.
event type 5004:
The Flagstons:
parents: Hi Lois 
sons: Chip Ditto 
daughters: Dot Trixie 
cats: 
dogs: Dawg 
computers: 
others: 
event type 5005:
The Flintstones:
fatherFlag: 1
motherFlag: 1
sonFlag: 0
daughterFlag: 1
catFlag: 0
dogFlag: 0
computerFlag: 0
otherFlag: 1
father: Fred 
mother: Wilma 
son: 
daughter: Pebbles 
cat: 
dog: 
computer: 
others: Dino 
event type 5006:
This message should be followed by 3 extra strings.
Extra string 1
Extra string 2
Extra string 3

event type 812611159:
Test of specifying event_type as string in template.
If you see this line, we probably passed.
